### ISSUE TITLE
[Impeller] Better utilization of display list recording format in Impeller.

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#142054]
Note: This is a somewhat ambigious title for a longer open ended project.

### Background

Flutter records canvas operations (draw rect, save layer) on the UI thread and plays these operations back on the raster thread. These operations are recorded in the display list format (See https://github.com/flutter/engine/tree/main/display_list ). To support the Skia -> Impeller transition, the format is _mostly_ backend agnostic and is converted to appropriate Skia/Impeller objects on the fly during the raster workload. For Impeller at least, this conversion can actually be quite costly. Furthermore, we have ended up with two implementations of various raster tasks such as bounds compuation and opacity peephole, which is not sustainable long term.

For reference, here are some traces of display list -> impeller conversions on a moderately complex project. The Impeller API requires multiple heap allocations per drawing command (entity, color source, geometry) which adds some extra inefficiency. We've noticed on android that allocation / deallocation is much slower on iOS.

![image](https://github.com/flutter/flutter/assets/8975114/7b98bae7-9dc8-4925-98e5-29e1ae2d9c87)

If we zoom in on this we can see its due to the excessive heap allocations in the Impeller API.

![image](https://github.com/flutter/flutter/assets/8975114/7d349a2e-97db-47bd-8fcf-a1bf85810ce3)

### Overview

At a high level, there are two problems with the display_list -> Impeller interop that need to be solved:

1. (easy) DisplayList itself contains mostly Skia objects.

For the most part this isn't _really_ a problem. The conversion of types like SkRect/SkPoint/SkColor don't appear in any of our traces or profiles. The conversion of SkPaths and TextBlobs are extremely expensive though, the latter has already been moved into display list via some hacks that I added. We also rely on the SkPath for correct/fast bounds computation (Impeller paths have slow bounds computations) and for the convexity check that is critical for tessellation performance.

2. (harder) EntityPass references display list + offsets instead of constructing new rendering tree

The most expensive part of the interop usually has to do with Impeller heap allocating a geometry/color source/entity for every single drawing command in the display list. I think the best approach involves changing entity pass itself to traverse the display list, and giving display list the appropriate APIs. In theory, this would allow almost no additional allocation in entity pass.

3. (cleanup) Impeller should be defering to display list bounds compuation.

Rather than try to fix Impeller's coverage calculations, we should have display list provide local/global bounds for all drawing commands that it has computed them for. We can then have a single means of computing coverage that we can ensure is correct.

Long term, we can provide display_list as the API for interacting with Flutter and delete aiks.

### Alternatives Considered

* Fix the Impeller API to perform fewer heap allocations.

We could flatten the impeller API into a display list like format. Why have two display list formats though?

* Make the Impeller API the recording format / remove display list.

@dnfield tried this in:  ___ . Our evaluation at the time was that it would take a lot of work to fix all of the bugs in Impeller's APIs to bring it up to display list quality. Another realization was that to achieve the performance we'd want, we'd need to flatten the recording structure into something that is display list like anyway.

**Comment by flar** [comment]
How about have Impeller Dispatcher dispatch directly into rendering commands rather than record things into intermediate entity passes. Right now, aiks Canvas is a recording format. Why is it involved in the final rendering phase?

Look at it this way. Skia has SkCanvas as the front end. That front end *can* be used to record operations for later. But that front end *can also* be used to just create commands for the GPU directly without any recording.

But with Impeller, it only has one Canvas and that Canvas essentially records into an Impeller "recording format" which is then immediately traversed to produce actual rendering commands. Why isn't there an impeller Canvas implementation that goes directly to the rendering commands without requiring a second traversal?

**Comment by bdero** [comment]
Yeah it feels like the right time to have an open discussion about this.

> I think the best approach involves changing entity pass itself to traverse the display list

I think standardizing around DisplayList and retiring the Aiks altogether is a good longterm goal (and is what @chinmaygarde was moving towards before we took off and started proving out Impeller really fast).

@flar 
> How about have Impeller Dispatcher dispatch directly into rendering commands rather than record things into intermediate entity passes.

The major issue with this used to be that we needed to count the number of subpasses that a sequence of commands will produce (which is backend-specific) in order to make efficient decisions about storing the multisample stencil. But that problem is largely gone now that we lazily replay clips as needed instead of ever storing rasterized clip state.

I'm also of the opinion that the subpass collapse optimization should happen before/while recording to the DisplayList. Lazy-able optimizations like the clear color optimization or stringing paths together can be done on either side of the fence.

One detail that would need to be dealt with here is that the `impeller::Picture` concept ceases to exist in this world, which seems fine. In this model, the DisplayList becomes the picture.

**Comment by flar** [comment]
Here is a list of Sk types used in the DisplayList records, as determined using `tr` to separate out words and grepping for `Sk`:

These are trivially castable by reference (the Impeller and Skia objects are identical simple structs), we could just switch the DL records to use the Impeller variants for simplicity and isolation from Skia source base without any performance issues:
- `SkIRect`
- `SkPoint`
- `SkRect`
- `SkScalar`

These are simple structs that Impeller doesn't have, but we could easily create a duplicate for and they could be trivially castable for the Skia backend similar to `SkRect`/`SkPoint`/et al:
- `SkRRect`
- `SkRSXform` (Impeller just immediately ingests these into a `Matrix`, but we don't want an array of `Matrix` objects in drawAtlas as it would take 4x the memory, so we should create an impeller native variation of this struct for data typing in the DL dispatch calls)

This one is a whole ball of wax and we need to come up with a solution (perhaps similar to the `SkTextBlob` solution where we have 2 calls, one for Skia with `SkPath` and one for Impeller `with impeller::Path` and we choose which to use at runtime? Or perhaps have a format that Impeller can use directly and which can be trivially converted to `SkPath` for the legacy backend?):
- `SkPath`

These are not passed in the dispatch functions, instead we pass in Nx`SkScalar` values that represent their data, but we could save a struct allocation if we passed them as a `Matrix` or similar object. This issue also intersects with a mechanism I'd like to develop for Impeller which is a tagged `Matrix` that optimizes the math when the `Matrix` is sparse (`impeller::Transform`?):
- `SkM44`
- `SkMatrix`

This object we still pass in one of the dispatch methods, but there is an equivalent version that passes an Impeller structure so we never need to convert it:
- SkTextBlob

**Comment by bdero** [comment]
> The biggest cost in conversion is the heap allocation of impeller:Contents/impeller::ColorSource et cetera

If this is the biggest concern, a much easier short term solution would be a per-Canvas arena.

Not saying we should definitely do this, but I quickly jotted down a design in case it's the _overwhelming_ issue you're seeing when profiling right now: https://github.com/flutter/flutter/issues/142191

> I think an initial goal would be allowing EntityPass to treat a display list as the set of high level rendering commands to iterate over.

I don't think this is distinct from what @flar is suggesting in [this comment](https://github.com/flutter/flutter/issues/142054#issuecomment-1906875242). The Impeller dispatcher _becomes_ `EntityPass`. The dispatcher just walks a DL buffer based on op sizes and does the big switch/dispatch that we have to do either way. I think the only actual difference would be that we'd do one extra vtable deref per draw, but even that can go away if/when we turn down Skia.

**Comment by jonahwilliams** [comment]
> If this is the biggest concern, a much easier short term solution would be a per-Canvas arena.

That is where I disagree. The per-canvas arena doesn't get us closer to the long term state we want, and arena allocation with shared ptrs its going to be a PTA unless we refactor entities substantially.

Right, the dispatcher becomes entity pass, or a similar API with some additional state tracking as needed. The skia conversions can happen on the fly, and then a subsequent project can strip those out gradually.

**Comment by flar** [comment]
> Right, the dispatcher becomes entity pass, or a similar API with some additional state tracking as needed. The skia conversions can happen on the fly, and then a subsequent project can strip those out gradually.

It would be nice to see a list of the kind of additional information we would need about op records in the DisplayList. DL computes bounds on everything, but then either throws out the individual answers and accumulates just a bounds, or it keeps them, but inside a DlRTree data structure that isn't readily accessible. Would it help if those were available? Any other info? Perhaps be able to provide the "current transform" and "clip bounds" for each op rather than the dispatchee having to do the record keeping itself?

**Comment by jonahwilliams** [comment]
Definitely the per command bounds/save layer bounds would be great to have. Impeller ends up re-computing these bounds, which can be quite slow for save layers that have a lot of entites - though your rect patch made it faster too.

**Comment by bdero** [comment]
> its going to be a PTA unless we refactor entities substantially.

@jonahwilliams Actually the entities part would be trivial. And it should be -- we're just lengthening the lifetime of some stuff. The diff in `EntityPass` and the `Contents` hierarchy would literally just be replacing some `std::shared_ptr` fields with `arena_ptr`, and adding a param to `Contents::Render` with a ref to the allocator (for the few places where we generate `Contents` on the fly like `AnonymousContents` in filters and `FramebufferBlendContents` in `EntityPass`).

Squishing `EntityPass`, `Aiks`, and `DlDispatcher` together and unraveling `Contents` is probably a multi-quarter effort at this point. The arena would probably be a few days of focused work, with almost all of the time spent on changes to `impeller::Canvas` and its usage around the engine. So might be a band-aid worth considering if allocation is really our biggest issue right now.

**Comment by bdero** [comment]
Some quick notes:

Today we're in a situation where we do the following full conversion passes:
1. **UI thread:** Record everything to a tree of **DisplayLists**.
2. **Raster thread:** Walk the tree of DisplayLists and convert all types to a tree of EntityPasses.
   - The EntityPass tree 1-1 corresponds with SaveLayers and almost always manages one texture at a time.
3. **Raster thread:** Walk the tree of **EntityPasses** and record the commands into a flattened list of Impeller **CommandBuffers**, and submit them.

The goal is to combine 2 and 3 into the same step, so that we have all of the context to record directly to **CommandBuffers**.

Before this can be viable, some subproblems that need to be solved:
- Any DrawPicture use cases that may exist today in the engine with Impeller need to be handled while building the DisplayList. Once Aiks `DrawPicture` is removed entirely, we've solved this subproblem.
- Before creating an offscreen RenderPass for a SaveLayer, we need a way to compute the minimum screen space bounds of the framebuffer we'll be rendering to. In other words we need, `EntityPass::GetSubpassCoverage`. This _must_ incorporate the coverage of all SaveLayers and their supplied filters, recursively. I strongly recommend we implement this API in the DisplayList itself.
- The opacity peephole optimization requires peeking ahead N number of operations in the DisplayList. So either:
  a. The DisplayList dispatching mechanism provides a way to peek future ops, or
  b. we implement this optimization above the DisplayList.

**Comment by jonahwilliams** [comment]
@flar I've assigned this to you for now.

**Comment by flar** [comment]
> * The opacity peephole optimization requires peeking ahead N number of operations in the DisplayList. So either:
>   a. The DisplayList dispatching mechanism provides a way to peek future ops, or
>   b. we implement this optimization while converting the layer tree to the DisplayList instead.

See https://github.com/flutter/engine/blob/500ae6c4d947879da2a879d8b2f1a4e7f4b41432/display_list/display_list.h#L178

**Comment by flar** [comment]
> * Before creating an offscreen RenderPass for a SaveLayer, we need a way to compute the minimum screen space bounds of the framebuffer we'll be rendering to. In other words we need, `EntityPass::GetSubpassCoverage`. This _must_ incorporate the coverage of all SaveLayers and their supplied filters, recursively. I strongly recommend we implement this API in the DisplayList itself.

Just to make sure we are on the same page, sometimes the caller calls `SaveLayer` with a bounds rectangle and the DL assumes that that rectangle will be honored even though Skia has always documented it as an optional hint that they might ignore. In the past, they've actually just used it as the bounds of the sub-layer (not the results of the saveLayer, but the size of the layer into which its content is rendered). It looks like `impeller::Canvas` forwards these bounds to the sub-pass so I think that happens here.

But, the caller doesn't have to supply a bounds rectangle in which case impeller wants to be able to compute the actual bounds to use. In the past, we've just forwarded this "lack of" bounds to Skia and let them compute the bounds because that is what happens when you record a Skia picture, but we actually do compute the bounds of the content of the layer because we have to compute the bounds of the DL. So, the value is available, we've computed it in DL, but we've already recorded the SL as a no-bounds SL in the stream, so we have no place to put it. We could probably solve this by simply recording all SL as SL-with-bounds, but marking the record as "needing bounds". Then when we hit the associated restore we can just back-fill the bounds from what we've calculated.

So, this item is likely already half-implemented and could be very easily fully-implemented if I understand everything correctly...

**Comment by flar** [comment]
There are a couple of corner cases when thinking about the DL saveLayer always providing the bounds of the content:

- Right now it passes along what the recording process told it, which I think should continue.
- If it calculates the bounds, normal rendering will tend to have well defined bounds which it could supply.
- Sometimes, though, the content of a layer (the **input** bounds) will be unbounded if there is an unclipped drawColor or drawPaint, or any other case that floods the output surface, such as some filters that turn transparency into non-transparency. In that case it will look for an enclosing clip to limit the **output** bounds of the DL, but it won't have a good answer for the input bounds. Luckily in that case since it is only tasked with computing the output bounds in the first place, it just ignores the lack of a "good answer" for the content bounds and instead accumulates the (clipped) output bounds for the DL/enclosing layer bounds. We'd need to solve that issue. If there are no filters on the saveLayer then the content bounds could just be the enclosing clip bounds. If there are filters then we'd need to do a reverse bounds on the clip to get the proper content bounds.
- If the saveLayer in question is at the root of the DL then there may be no enclosing clip within the DL against which to perform the above calculations. However, when this DL is played back there definitely will be an enclosing clip, either coming from the parent DL, or coming eventually from the size of the rendering target. In that case we might not be able to supply the bounds when recording the DL, but the code being dispatched to might be able to come up with one from its environment. It won't have to do a bounds analysis of the content to do so, though, it will just have to look at any enclosing clips or the render target size (and perhaps do an inverse-filter-bounds adjustment if there is a filter at play).

It should be easy to provide some content bounds in all saveLayer calls for the simple case of bounded content - something that isn't being done at the moment.

With a small amount of work, it should be easy to provide clip-relative bounds for the unbounded content case, by doing a reverse filter bounds operation on the enclosing clip - if there is one.

In the case of unbounded content with no enclosing clip, though, the unbounded classification will have to be passed onto and be the burden of the dispatcher.

**Comment by jonahwilliams** [comment]
Here is an idea about an intermediate step that might be less disruptive than EntityPass2::ElectricBoogalo.

Right now we have a multi-step process:

```c++
DisplayList display_list.
DLDispatcher dispatcher;
display_list->dispatch(dispatcher);
auto picture = dispatcher.EndRecording();
aiks_context->Render(picture)
```

dispatch is where we end up copying the structure of the display list into Impeller concepts: ColorSource, Geometry, Entity. this is also where all of the heap allocations happen.

We could create an alternative dispatch interface, instead of:

```c++
void DisplayList::Dispatch(DlOpReceiver& receiver,
                           uint8_t* ptr,
                           uint8_t* end,
                           Culler& culler) const {
...
  while (ptr < end) {
    auto op = reinterpret_cast<const DLOp*>(ptr);
    switch (op.type) {
      // Individual dispatch methods.
    }
```

We create an interface that hands pointers into the display list to a dispatcher.

```c++
void DisplayList::Dispatch(DlOpReceiver2& receiver,
                           uint8_t* ptr,
                           uint8_t* end,
                           Culler& culler) const {
...
  while (ptr < end) {
    auto op = reinterpret_cast<const DLOp*>(ptr);
    receiver.storeOp(op);
    ...
```

(Ignore the Paint methods for now, pretend everything is a draw call like drawRect, saveLayer) et cetera. Now rather than each entity containing a geomtry/color source, it just has a ptr to the display list and some metadata about the specific draw:

```c++
class DLContents : public Contents {

  std::optional<Rect> GetCoverage(..) {
    return display_list->queryBounds(data);
  }
  
  bool Render(const ContentContext& renderer,
                      const Entity& entity,
                      RenderPass& pass) {
    Geometry& geometry = createGeometry(data, display_list);
    ColorSource& color_source = createColorSource(data, display_list);
    color_source.SetGeometry(geometry);
    return color_source.Render(...);
  }
  const DLOp* data;
  const DisplayList* display_list;
}

```

Something like that. We still create the ColorSource/Geometry, but we only do it on the fly - this means we can probably structure the code so that we don't have to malloc the data. The overall picture could store a shared_ptr to the display list to keep it alive appropriately as well.

**Comment by jonahwilliams** [comment]
@flar  I think as long as we could get the local bounds of the individual draws, we wouldn't need to solve the problem of saveLayer bounds until layer.

**Comment by bdero** [comment]
@flar (In response to private messages) As promised, here's the full list of look-aheads that Aiks/EntityPass currently does -- each of these are essential features/optimizations that eventually need to be replaced:
1. Assign a depth value to all draw ops. For non-clips, assign incremental "depth" values (where greater depth means "closer to the camera"). For clips, assign the max depth value of any non-clip draw op that the clip affects.
2. **For optimizing texture sizes:** Compute (unfiltered) conservative bounds of a SaveLayer (i.e. `GetSubpassCoverage`). If these bounds are local space, that's fine since we'll be tracking the CTM anyhow.
3. **For minimizing glyph atlas uploads:** Before rendering anything, iterate through all text frames in the scene to populate glyph atlases in advance.
4. When a SaveLayer with a non-destructive blend begins with N non-overlapping draw ops, we render those N draw ops directly into the parent layer instead of allocating a new texture.
5. **Optimize away full texture fills:** Before creating a texture for a SaveLayer, we walk through draw operations at the beginning and blend them into the texture's clear color on the CPU, then skip over those operations. It would be great to remove this complexity, but technically we can accomplish this without look-ahead by lazily delaying `RenderPass` creation until we run into a draw op that doesn't perform a full texture fill.

**Comment by bdero** [comment]
@jonahwilliams
> Here is an idea about an intermediate step that might be less disruptive than EntityPass2::ElectricBoogalo.

Correct me if I'm wrong, but my understanding of this idea is that we'd start by making a new kind of `flutter::DlOpReceiver` and replace the `ColorSourceContents`/`Geometry`/`Aiks`/`DlDispatcher` code without making any changes to `EntityPass`? This way, we would still be individually boxing `Contents`, but not `Geometry` or `ColorSources`.

> We could create an alternative dispatch interface ... We create an interface that hands pointers into the display list to a dispatcher

> Now rather than each entity containing a geomtry/color source, it just has a ptr to the display list and some metadata about the specific draw

With your example in mind:
> ```cpp
>     Geometry& geometry = createGeometry(data, display_list);
>     ColorSource& color_source = createColorSource(data, display_list);
>     color_source.SetGeometry(geometry);
>     return color_source.Render(...);
> ```

Overall there's some goodness here, but I don't see how some aspects of this can work. Thoughts:
1. Given the proposed generic `storeOp` dispatch, how did you decide to create a `DLContents` in the first place? Wouldn't we'd need to disambiguate if it's one of a handful of ops upfront, lest we wind up allocating mostly no-op `Contents`?
2. It's unclear how we're resolving both the geometry and the color source from the given `DlOp`. Presumably we already know the `DlOp` is going to be something like `DrawRect` already given we're rendering in `DLContents`, so pulling the geometry out makes sense. But for the color source we'd need to backwards search the DL to find the most recent color source. It would be much easier to just use the original `DlOpReceiver` interface and track/update a `DlColorSource` reference as we receive ops to do so over time.
3. To clarify -- given how the example is written, `createGeometry` and `createColorSource` _must be_ allocating, right? We just haven't defined how/where yet. Granted, it's deferred until render time and so the lifetime requirement of the allocation is super short, which is obviously advantageous.
4. Why have a separate `ColorSource` piece at all? Why not make new `Contents` that implement each color source conversion in `Render` directly? Not to mention all the oddball draws that don't cleanly fit into the `Geometry`+`ColorSource` pattern like `DrawAtlas` and `DrawVertices`. Later on when we merge `EntityPass` with the new `DlDispatcher` and the need for `Contents` boxing melts away, we can just store the derived `DlColorSourceContents` in a union field on the paint state directly and dynamic dispatch the `Render` method, for example.

I think making a new `ColorSourceContents` hierarchy that uses DL refs would be a solid step towards our end goal here, but deferring the op dispatch sounds like pure pain for no gainz.

> The overall picture could store a shared_ptr to the display list to keep it alive appropriately as well.

Alternatively, we could just nip this in the bud early. The new dispatcher can just not allow an `impeller::Picture` to be created. We can force the two stages together and disallow the possibility of creating or holding on to an `impeller::Picture`. This protection is trivial iff passing DL pointers to `EntityPass` is actually safe for all our current usages:
```cpp
bool AiksContext::RenderDisplayList(const DisplayList& display_list,
                                    Rect cull_rect,
                                    RenderTarget& render_target,
                                    bool reset_host_buffer) {
  // ... transients reset, etc ...

  return DlDispatcher2::Render(*content_context_, render_target,
                               display_list, cull_rect);
}
```

```cpp
class DlDispatcher2 final : public flutter::DlOpReceiver {
 private:
  DlDispatcher2();  // Must use `DlDispatcher2::Render`.

 public:
  /// Create the dispatcher, dispatch, and render. Do NOT supply an interface
  /// to create an `impeller::Picture`
  static bool DlDispatcher2::Render(ContentContext& content_context,
                                    const RenderTarget& render_target
                                    const DisplayList& display_list,
                                    Rect cull_rect) {
    impeller::DlDispatcher2 dispatcher_v2(cull_rect);
    display_list.dispatch(dispatcher_v2);

    impeller::Picture picture = canvas.EndRecordingAsPicture();
    if (picture.pass) {
      return picture.pass->Render(content_context, render_target);
    }
    return true;
  }
  // ...
};
```

**Comment by flar** [comment]
We are still actively working on having a more direct path from Flutter `ui` through `DisplayList` to Impeller's rendering code. PRs are being landed on a weekly, sometimes daily, basis to make that path more direct and efficient, but we aren't linking them to this issue necessarily.

**Comment by jonahwilliams** [comment]
Tons of status updates bird, you just need to look in the linked PRs!
